src/mlpack/tests/ne_test.cpp

```diff
@@ -0,0 +1,425 @@
+#include <mlpack/methods/ne/utils.hpp>
+#include <mlpack/methods/ne/parameters.hpp>
+#include <mlpack/methods/ne/tasks.hpp>
+#include <mlpack/methods/ne/link_gene.hpp>
+#include <mlpack/methods/ne/neuron_gene.hpp>
+#include <mlpack/methods/ne/genome.hpp>
+#include <mlpack/methods/ne/species.hpp>
+#include <mlpack/methods/ne/cne.hpp>
+#include <mlpack/methods/ne/neat.hpp>
+
+#include <boost/test/unit_test.hpp>
+#include "old_boost_test_definitions.hpp"
+
+using namespace mlpack;
+using namespace mlpack::ne;
+
+BOOST_AUTO_TEST_SUITE(NETest);
+
+/**
+ * Test LinkGene class.
+ */
+BOOST_AUTO_TEST_CASE(NELinkGeneTest)
+{
+  // Create a link gene.
+  LinkGene linkGene(1, 2, 0, 10, true);
+
+  // Test parametric constructor and access functions.
+  BOOST_REQUIRE_EQUAL(linkGene.FromNeuronId(), 1);
+  BOOST_REQUIRE_EQUAL(linkGene.ToNeuronId(), 2);
+  BOOST_REQUIRE_EQUAL(linkGene.InnovationId(), 0);
+  BOOST_REQUIRE_EQUAL(linkGene.Weight(), 10);
+
+  // Test set function.
+  linkGene.Weight(20);
+  BOOST_REQUIRE_EQUAL(linkGene.Weight(), 20);
+
+  // Test copy constructor.
+  LinkGene linkGene2(linkGene);
+  BOOST_REQUIRE_EQUAL(linkGene2.FromNeuronId(), 1);
+
+  // Test operator =.
+  LinkGene linkGene3(1, 3, 1, 7, true);
+  linkGene = linkGene3;
+  BOOST_REQUIRE_EQUAL(linkGene3.InnovationId(), 1);
+
+}
+
+/**
+ * Test NeuronGene class.
+ */
+BOOST_AUTO_TEST_CASE(NENeuronGeneTest)
+{
+  // Create a neuron gene.
+  NeuronGene neuronGene(1, INPUT, SIGMOID, 0, 0, 0);
+
+  // Test parametric constructor and access functions.
+  BOOST_REQUIRE_EQUAL(neuronGene.Id(), 1);
+  BOOST_REQUIRE_EQUAL(neuronGene.Type(), INPUT);
+  BOOST_REQUIRE_EQUAL(neuronGene.ActFuncType(), SIGMOID);
+  BOOST_REQUIRE_EQUAL(neuronGene.Input(), 0);
+  BOOST_REQUIRE_EQUAL(neuronGene.Activation(), 0);
+
+  // Test set function
+  neuronGene.ActFuncType(RELU);
+  BOOST_REQUIRE_EQUAL(neuronGene.ActFuncType(), RELU);
+
+  // Test copy function.
+  NeuronGene neuronGene2(neuronGene);
+  BOOST_REQUIRE_EQUAL(neuronGene2.Id(), 1);
+
+  // Test operator =.
+  NeuronGene neuronGene3(11, INPUT, SIGMOID, 0, 0, 0);
+  neuronGene = neuronGene3;
+  BOOST_REQUIRE_EQUAL(neuronGene.Id(), 11);
+
+}
+
+/**
+ * Test Genome class.
+ */
+BOOST_AUTO_TEST_CASE(NEGenomeTest)
+{
+  mlpack::math::RandomSeed(1);
+
+  // Construct seed genome for xor task.
+  ssize_t id = 0;
+  ssize_t numInput = 3;
+  ssize_t numOutput = 1;
+  double fitness = -1;
+  std::vector<NeuronGene> neuronGenes;
+  std::vector<LinkGene> linkGenes;
+
+  NeuronGene inputGene1(0, INPUT, LINEAR, 0, 0, 0);
+  NeuronGene inputGene2(1, INPUT, LINEAR, 0, 0, 0);
+  NeuronGene biasGene(2, BIAS, LINEAR, 0, 0, 0);
+  NeuronGene outputGene(3, OUTPUT, SIGMOID, 1, 0, 0);
+  NeuronGene hiddenGene(4, HIDDEN, SIGMOID, 0.5, 0, 0);
+
+  neuronGenes.push_back(inputGene1);
+  neuronGenes.push_back(inputGene2);
+  neuronGenes.push_back(biasGene);
+  neuronGenes.push_back(outputGene);
+  neuronGenes.push_back(hiddenGene);
+
+  LinkGene link1(0, 3, 0, 1, true);
```

Comment:
Oh yeah! Right, revised it!

2016-07-16 11:15 GMT-06:00 Marcus Edel <notifications@github.com>:

> In src/mlpack/tests/ne_test.cpp
> <https://github.com/mlpack/mlpack/pull/686#discussion_r71069041>:
>
> > +  std::vector<NeuronGene> neuronGenes;
> > +  std::vector<LinkGene> linkGenes;
> > +
> > +  NeuronGene inputGene1(0, INPUT, LINEAR, 0, 0, 0);
> > +  NeuronGene inputGene2(1, INPUT, LINEAR, 0, 0, 0);
> > +  NeuronGene biasGene(2, BIAS, LINEAR, 0, 0, 0);
> > +  NeuronGene outputGene(3, OUTPUT, SIGMOID, 1, 0, 0);
> > +  NeuronGene hiddenGene(4, HIDDEN, SIGMOID, 0.5, 0, 0);
> > +
> > +  neuronGenes.push_back(inputGene1);
> > +  neuronGenes.push_back(inputGene2);
> > +  neuronGenes.push_back(biasGene);
> > +  neuronGenes.push_back(outputGene);
> > +  neuronGenes.push_back(hiddenGene);
> > +
> > +  LinkGene link1(0, 3, 0, 1, true);
>
> Maybe I missed something but I think each link should have a different
> innovation number.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mlpack/mlpack/pull/686/files/fc982b9406838eb9697000481e816165702e8ba9#r71069041>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD9tIX9w--WKopYPE34NivvWMlbI_phgks5qWRGlgaJpZM4IwJa6>
> .
>